sdks/python/apache_beam/runners/interactive/display/pcoll_visualization.py

```diff
@@ -0,0 +1,279 @@
+    # Disables the verbose logging from timeloop.
+    logging.getLogger('timeloop').disabled = True
+    tl = Timeloop()
+
+    def dynamic_plotting(pcoll, pv, tl):
+      @tl.job(interval=timedelta(seconds=dynamic_plotting_interval))
+      def continuous_update_display():  # pylint: disable=unused-variable
+        # Always creates a new PCollVisualization instance when the
+        # PCollection materialization is being updated and dynamic
+        # plotting is in-process.
+        updated_pv = PCollVisualization(pcoll)
+        updated_pv.display_facets(updating_pv=pv)
+        if ie.current_env().is_terminated(pcoll.pipeline):
+          try:
+            tl.stop()
+          except RuntimeError:
+            # The job can only be stopped once. Ignore excessive stops.
+            pass
+
+      tl.start()
+      return tl
+
+    return dynamic_plotting(pcoll, pv, tl)
+  return None
+
+
+class PCollVisualization(object):
+  """A visualization of a PCollection.
+
+  The class relies on creating a PipelineInstrument w/o actual instrument to
+  access current interactive environment for materialized PCollection data at
+  the moment of self instantiation through cache. It utilizes Facets, pandas
+  DataFrame and jQuery DataTable to visualize the PCollection data.
+  """
+
+  def __init__(self, pcoll):
+    assert isinstance(pcoll, pvalue.PCollection), (
+        'pcoll should be apache_beam.pvalue.PCollection')
+    self._pcoll = pcoll
+    # This allows us to access cache key and other meta data about the pipeline
+    # whether it's the pipeline defined in user code or a copy of that pipeline.
+    # Thus, this module doesn't need any other user input but the PCollection
+    # variable to be visualized. It then automatically figures out the pipeline
+    # definition, materialized data and the pipeline result for the execution
+    # even if the user never assigned or waited the result explicitly.
+    # With only the constructor of PipelineInstrument, any interactivity related
+    # pre-process or instrument is not triggered for performance concerns.
+    self._pin = instr.PipelineInstrument(pcoll.pipeline)
+    self._cache_key = self._pin.cache_key(self._pcoll)
+    self._dive_display_id = 'facets_dive_{}_{}'.format(self._cache_key,
+                                                       id(self))
+    self._overview_display_id = 'facets_overview_{}_{}'.format(self._cache_key,
+                                                               id(self))
+    self._df_display_id = 'df_{}_{}'.format(self._cache_key, id(self))
+
+  def display_facets(self, updating_pv=None):
+    """Displays the visualization through IPython.
+
+    The visualization has 3 parts: facets-dive, facets-overview and paginated
+    data table. Each part is assigned an auto-generated unique display id
+    (the uniqueness is guaranteed throughout the lifespan of the PCollection
+    variable). If a PCollVisualization instance is provided as updating_pv,
```

Comment:
You might document updating_pv and extract out that part of the single paragraph here.

Reply:
Done.